### PR TITLE
Pipeline client 6.0 added to support template parameters for pipelines runs

### PR DESCRIPTION
**Pull request by roshan-sy**
Please make sure the code is following contribution guidelines in CONTRIBUTING.md

 - [ ] : This PR has a corresponding issue open in the Repository.
 - [ ] : Approach is signed off on the issue.

Fixes: #1110 

1. Added pipeline client 6.0,
2. Added template parameters to the new run command
3. Command uses `v6_0 - PipelinesClient` when the "parameter" subcommand is provided.
4. For the rest use cases it uses `v5_0 - BuildClient`

### CODE REVIEW

File: azure-devops/azext_devops/dev/pipelines/pipeline.py [hunk 1]
```diff
@@ -123,8 +130,36 @@ def pipeline_run(id=None, branch=None, commit_id=None, name=None, open=False, va
...
     if id is None:
         id = get_definition_id_from_name(name, client, project, folder_path)
+
+    if parameters:
+        client = get_new_pipeline_client_v60(organization)
+
+        repositories = {"self": RepositoryResourceParameters(ref_name=branch, version=commit_id)}
+        resources = RunResourcesParameters(repositories=repositories)
+
+        template_parameters = ast.literal_eval(parameters)
```
Comment (gauravsaralMs) [inline azure-devops/azext_devops/dev/pipelines/pipeline.py:143]: what does this function do ?

I am assuming from a user point of view i need to enter similar values for parameters and variables...
then why to have seperate code to handle those ?

File: azure-devops/azext_devops/dev/pipelines/pipeline.py [hunk 2]
```diff
@@ -123,8 +130,36 @@ def pipeline_run(id=None, branch=None, commit_id=None, name=None, open=False, va
...
+
+        param_variables = None
+        if variables:
+            param_variables = {}
+            for variable in variables:
+                separator_pos = variable.find('=')
+                if separator_pos >= 0:
+                    param_variables[variable[:separator_pos]] = {"value": variable[separator_pos + 1:]}
+                else:
+                    raise ValueError('The --variables argument should consist of space separated "name=value" pairs.')
```
Comment (gauravsaralMs) [inline azure-devops/azext_devops/dev/pipelines/pipeline.py:153]: similar logic seems to exist below as well, can we refactor code to have the logic only in one place?

File: azure-devops/azext_devops/dev/pipelines/pipeline.py [hunk 3]
```diff
@@ -96,8 +101,31 @@ def pipeline_show(id=None, name=None, open=False, organization=None, project=Non
...
+                if as_dict:
+                    param_variables[variable[:separator_pos]] = {"value": variable[separator_pos + 1:]}
+                else:
+                    param_variables[variable[:separator_pos]] = variable[separator_pos + 1:]
+            else:
+                raise ValueError('The --variables argument should consist of space separated "name=value" pairs.')
+    return param_variables
+
+
+def convert_param(parameters):
```
Comment (gauravsaralMs) [inline azure-devops/azext_devops/dev/pipelines/pipeline.py:120]: as per def both variable and param are type dict in the class definition, then why this special handling for param ?

https://github.com/microsoft/azure-devops-python-api/blob/451cade4c475482792cbe9e522c1fee32393139e/azure-devops/azure/devops/v6_0/pipelines/models.py#L364-L367

**Review by gauravsaralMs — CHANGES_REQUESTED** [review]

File: azure-devops/azext_devops/dev/pipelines/pipeline.py [hunk 4]
```diff
@@ -96,8 +100,24 @@ def pipeline_show(id=None, name=None, open=False, organization=None, project=Non
...
+        param_variables = {}
+        for variable in variables:
+            separator_pos = variable.find('=')
+            if separator_pos >= 0:
+                if as_dict:
+                    param_variables[variable[:separator_pos]] = {"value": variable[separator_pos + 1:]}
+                else:
+                    param_variables[variable[:separator_pos]] = variable[separator_pos + 1:]
+            else:
+                raise ValueError('The --variables argument should consist of space separated "name=value" pairs.')
```
Comment (gauravsaralMs) [inline azure-devops/azext_devops/dev/pipelines/pipeline.py:115]: error should be different in case of different things

**Review by gauravsaralMs — CHANGES_REQUESTED** [review]
please 
- update PR title to show real intent of this PR
- add more details in the PR like why this change, what is the impact of this change

File: azure-devops/azext_devops/dev/pipelines/pipeline.py [hunk 5]
```diff
@@ -96,8 +100,25 @@ def pipeline_show(id=None, name=None, open=False, organization=None, project=Non
...
+        for variable in variables:
+            separator_pos = variable.find('=')
+            if separator_pos >= 0:
+                if as_dict:
+                    param_variables[variable[:separator_pos]] = {"value": variable[separator_pos + 1:]}
+                else:
+                    param_variables[variable[:separator_pos]] = variable[separator_pos + 1:]
+            else:
+                raise ValueError(
+                    f'The --{argument} argument should consist of space separated "name=value" pairs.')
```
Comment (gauravsaralMs) [inline azure-devops/azext_devops/dev/pipelines/pipeline.py:116]: can you share the screenshot on how this will look to customer ?

Comment (gauravsaralMs) [reply azure-devops/azext_devops/dev/pipelines/pipeline.py:116]: If it is going to look just like 
```
The --{argument} argument should consist of space separated "name=value" pairs.
```
then how would it help customers ?

File: azure-devops/azext_devops/dev/pipelines/pipeline.py [hunk 6]
```diff
@@ -110,6 +131,8 @@ def pipeline_run(id=None, branch=None, commit_id=None, name=None, open=False, va
     :type folder_path: str
     :param variables: Space separated "name=value" pairs for the variables you would like to set.
     :type variables: [str]
+    :param parameters: Space separated "name=value" pairs for the variables you would like to set.
```
Comment (gauravsaralMs) [inline azure-devops/azext_devops/dev/pipelines/pipeline.py:134]: here the help text needs to be updated

File: azure-devops/azext_devops/dev/pipelines/pipeline.py [hunk 7]
```diff
@@ -123,20 +146,35 @@ def pipeline_run(id=None, branch=None, commit_id=None, name=None, open=False, va
         raise ValueError('Either the --id argument or the --name argument ' +
                          'must be supplied for this command.')
     client = get_build_client(organization)
+
     if id is None:
         id = get_definition_id_from_name(name, client, project, folder_path)
+
+    if parameters:
+        client = get_new_pipeline_client_v60(organization)
```
Comment (gauravsaralMs) [inline azure-devops/azext_devops/dev/pipelines/pipeline.py:154]: add some more logging which can tell which code path is being taken, if a customer reports with debug log then that will be helpful

File: azure-devops/azext_devops/dev/pipelines/pipeline.py (hunk 5 above)
Comment (roshan-sy) [reply azure-devops/azext_devops/dev/pipelines/pipeline.py:116]: image

File: azure-devops/azext_devops/dev/pipelines/pipeline.py (hunk 6 above)
Comment (roshan-sy) [reply azure-devops/azext_devops/dev/pipelines/pipeline.py:134]: Done

**Review by gauravsaralMs — APPROVED** [review]

**Pull request merged**